crates/service/tests/router_test.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::time::{Duration, SystemTime};
+
+use alloy::primitives::Address;
+use axum::http::Request;
+use axum_extra::headers::Header;
+use indexer_config::{BlockchainConfig, GraphNodeConfig, IndexerConfig, NonZeroGRT};
+use indexer_monitor::EscrowAccounts;
+use indexer_service_rs::{
+    service::{ServiceRouter, TapReceipt},
+    QueryBody,
+};
+use reqwest::{StatusCode, Url};
+use sqlx::PgPool;
+use test_assets::{create_signed_receipt, INDEXER_ALLOCATIONS, TAP_EIP712_DOMAIN};
+use tokio::sync::watch;
+use tower::ServiceExt;
+use wiremock::{
+    matchers::{method, path},
+    Mock, MockServer, ResponseTemplate,
+};
+
+#[sqlx::test(migrations = "../../migrations")]
+async fn full_integration_test(database: PgPool) {
+    let http_client = reqwest::Client::builder()
+        .tcp_nodelay(true)
+        .build()
+        .expect("Failed to init HTTP client");
+
+    let allocation = INDEXER_ALLOCATIONS
+        .values()
+        .collect::<Vec<_>>()
+        .pop()
+        .unwrap()
+        .clone();
+    let deployment = allocation.subgraph_deployment.id;
+
+    let mock_server = MockServer::start().await;
+
+    let mock = Mock::given(method("POST"))
+        .and(path(format!("/subgraphs/id/{deployment}")))
+        .respond_with(ResponseTemplate::new(200).set_body_raw(
+            r#"
+                {
+                    "data": {
+                        "graphNetwork": {
+                            "currentEpoch": 960
+                        }
+                    }
+                }
+                "#,
+            "application/json",
+        ));
+    mock_server.register(mock).await;
+
+    let (_escrow_tx, escrow_accounts) = watch::channel(EscrowAccounts::new(
+        test_assets::ESCROW_ACCOUNTS_BALANCES.clone(),
+        test_assets::ESCROW_ACCOUNTS_SENDERS_TO_SIGNERS.clone(),
+    ));
+    let (_dispute_tx, dispute_manager) = watch::channel(Address::ZERO);
+
+    let (_allocations_tx, allocations) = watch::channel(test_assets::INDEXER_ALLOCATIONS.clone());
+
+    let graph_node_url = Url::parse(&mock_server.uri()).unwrap();
+
+    let router = ServiceRouter::builder()
+        .database(database)
+        .domain_separator(TAP_EIP712_DOMAIN.clone())
+        .http_client(http_client)
+        .graph_node(leak(GraphNodeConfig {
```

Comment:
I am kinda concerned by the number of `leak`s that I see in this PR. Could you talk me through why we need these? Could we refactor such that we don't need `'static` lifetimes?

Reply:
I see you replaced some of them with `Arc` in your last commit, but there are a lot remaining.

Reply:
All of these are from configuration files. Previously we had just `&'static Config` but this was really hard to test since you need to mock a bunch of config values that are not used.

I decided to use just the references needed for the Router. I'll try to refactor those to use just the reference instead of static references.

The reason why this is not an issue is because at runtime we already have config static.

There are 2 differences between a leaked reference vs Arcs that live the whole program:
- Arcs have the overhead of cloning
- leaked objects never run the Drop trait

The only reason I changed Manager from static to Arc is because we need to run drop and close the database connections, this got clearer with tests.